shared-worker/firebase-worker.js

```diff
@@ -0,0 +1,54 @@
+importScripts('../firebase/firebase-app.js');
+importScripts('../firebase/firebase-firestore.js');
+var config = {
+  // your config here
+};
+
+/** Port used to communicate with the shared worker. */
+self.sharedWorkerPort = null;
+
+/** Whether this Firestore worker holds the lease. */
+self.hasPrimaryLease = false;
+
+/** The Firestore instance. */
+self.db = null;
+
+self.onmessage = function(e) {
+  // Setup message listeners after receiving the Shared Worker's port and
+  // request the lease.
+  if (e.data.sharedWorkerPort) {
+    self.sharedWorkerPort = e.data.sharedWorkerPort;
+    self.sharedWorkerPort.onmessage = (event) => handleLeaseEvent(event);
+    self.sharedWorkerPort.postMessage('requestingLease');
+  } 
+
+  // Release the primary lease in the shared worker if this worker owns the
+  // lease when this worker is being terminated.
+  if (e.data === 'terminate' && self.hasPrimaryLease) {
+    self.sharedWorkerPort.postMessage('releaseLease');
+  } 
+};
+
+async function initializeApp(takeLease) {
+  const app2 = firebase.initializeApp(config, "some-namez");
+  self.db = firebase.firestore(app2);
+  if (takeLease) {
+    await self.db.enablePersistence({experimentalForce: true});
+  } else {
+    await self.db.enablePersistence();
```

Comment:
This is likely going to fail (because another worker has the lease), right?  Or no?

It seems like this is the point where the developer needs to make a decision.  Do they give up (and show the user a message like "The app is open in another tab.  Please close this tab and switch to that one.")?  Or do they go ahead and enable Firestore, probably without persistence (in which case should we be calling enablePersistence() at all here)?.  If we end up calling enablePersistence() even when we don't expect it to succeed, we should probably still handle the result and if it *does* succeed we presumably need to notify the shared worker that we have the lease, right?

Reply:
I left it as a comment, since that will trigger the error that is thrown when trying to enable persistence when it's not available.